### ISSUE TITLE
2D images are distorted with opengl3 driver

### ISSUE THREAD

**Issue by Zughy** [luanti-org/luanti#15524]
### Luanti version

```true
5.11-dev, latest

-- BUILD_CLIENT: TRUE
-- BUILD_SERVER: FALSE
-- BUILD_UNITTESTS: TRUE
-- BUILD_BENCHMARKS: FALSE
-- BUILD_DOCUMENTATION: TRUE
-- RUN_IN_PLACE: 1
-- Using imported IrrlichtMt at subdirectory 'irr'
-- *** Building IrrlichtMt ***
-- Device: SDL
-- OpenGL: ON
-- OpenGL 3: ON
-- OpenGL ES 2: OFF
```

### Irrlicht device

SDL

### Operating system and version

Pop!_OS 22.04 LTS

### GPU model

GeForce GTX 1060

### Active renderer

4.6.0 NVIDIA 560.35.03 (what?)

### Summary

I noticed that with Soothing32, this happens (notice the 4th heart)  

![image](https://github.com/user-attachments/assets/d3a766ec-7e82-4dd6-b331-125c53eb04ff)

It also incorrectly renders the stamina of Block League, which is a perfect rectangle  

![image](https://github.com/user-attachments/assets/8f14f53e-3eb1-440e-9d79-e1ea677c3c5f)

comparison with 5.10

![image](https://github.com/user-attachments/assets/c40aa7b6-f2e5-45fd-b0ac-a4a2fa93f95f)

### Steps to reproduce

1. Download Soothing32 texture pack
2. Enable it
3. Open whatever devtest world

In alternative, join A.E.S. server

**Comment by appgurueu** [comment]
Very odd. It looks like it's rotating the stamina bar a very slight bit? Not sure what happens with the hearts.

Can you try to [bisect](https://www.metaltoad.com/blog/beginners-guide-git-bisect-process-elimination) this?

**Comment by Desour** [comment]
Related: #10733

**Comment by Zughy** [comment]
@appgurueu apparently the issue is not a commit in 5.11 (I bisected until 5.10) but `video_driver = opengl3`

**Comment by v-rob** [comment]
There are other distortions with the OpenGL3 driver. Notably, `draw2DImage` often has minor distortions, which I believe is due to the fact that it is rendered as two triangles (notice how the distortions follow the diagonal from the top left of the image to the bottom right).

| Legacy GL | GL 3 |
|---|---|
| ![image](https://github.com/user-attachments/assets/625e2ec5-6aad-4153-aeb1-c20256714535) | ![image](https://github.com/user-attachments/assets/010cad4b-fec6-4e59-8e71-a9beff0503fe) |

I suspect this is related to the other distortions, since the HUD uses `draw2DImageFilterScaled` to draw images, which in turn uses `draw2DImage`.

**Comment by sfan5** [comment]
> which I believe is due to the fact that it is rendered as two triangles

the legacy driver too uses triangles:
https://github.com/minetest/minetest/blob/f99a1a7c7c1a3d96332169978b63ffb10c56dbb9/irr/src/COpenGLDriver.cpp#L1100-L1113

**Comment by SmallJoker** [comment]
**AMD Polaris series card**, amdgpu driver from Mesa 24.3.0
No visual anomalies.
![luanti_7bf0b1f_amdgpu_polaris](https://github.com/user-attachments/assets/56ef206c-11a9-4c65-afbe-d7c90ec7c916)

**NVIDIA Pascal series card**, proprietary driver, 535.183.06
The first and 6th heart is dented.
![luanti_7bf0b1f_nvidia_pascal](https://github.com/user-attachments/assets/08c05beb-3a80-4b88-937d-450446941659)

**Comment by v-rob** [comment]
I am unable to reproduce the hearts issue exactly, but I do get the issue in general (like the `air.png` examples above) with AMD Radeon integrated graphics, so I don't think it's necessarily NVIDIA-exclusive (although they might just be different root issues).

@SmallJoker Could you check for anomalies using
```minetest.show_formspec("singleplayer", "test", "formspec_version[3]size[3,5]image[0,0;3,5;air.png]")```
The issue may not appear/be less apparent at certain window sizes.

**Comment by SmallJoker** [comment]
minetest.conf
```
video_driver = opengl3
#video_driver = opengl
opengl_debug = true
screen_w = 1000
screen_h = 800
```


| Card  | opengl | opengl3 |
|--|--|--|
| Polaris | ![amdgpu_polaris_opengl](https://github.com/user-attachments/assets/5495bae9-adaa-4b08-ac0b-964adb91250a) | ![amdgpu_polaris_opengl3](https://github.com/user-attachments/assets/abe3aca1-4fc3-4225-ae34-3dfd59a0f0eb) |
| Pascal | ![nvidia_opengl](https://github.com/user-attachments/assets/937d80e1-7b26-4f77-9bcc-fd8ecf1187d9) | ![nvidia_opengl3](https://github.com/user-attachments/assets/fa03e711-a978-41cc-bd4c-7b78bc4d53db) |
| Intel HD | ![intel_opengl](https://github.com/user-attachments/assets/1be59fc5-bdf0-467a-a040-1d6eefea2763) | ![intel_opengl3](https://github.com/user-attachments/assets/b92450f2-1df0-4ce6-b869-4a1b4c876df2) |

**Comment by SmallJoker** [comment]
While trying to implement `COpenGL3DriverBase::draw2DVertexPrimitiveList` I ran into this strange calculation:
https://github.com/minetest/minetest/blob/83bc27d99dacbb249b95387a9511d9b38830bf55/irr/src/OpenGL/Driver.cpp#L1005-L1019

Meanwhile on the older implementation does not scale by `getCurrentRenderTargetSize`:
https://github.com/minetest/minetest/blob/83bc27d99dacbb249b95387a9511d9b38830bf55/irr/src/COpenGLDriver.cpp#L1495-L1506

Might the viewport or the projection matrix be different in those two implementations? If so, that might explain erroneous distortions due to rounding issues.

**Comment by sfan5** [comment]
Those are normalized-device coordinate, see https://learnopengl.com/Getting-started/Coordinate-Systems.

the legacy driver uses a projection matrix (and some other parameters?) that adjusts for the RTT size:
https://github.com/minetest/minetest/blob/83bc27d99dacbb249b95387a9511d9b38830bf55/irr/src/COpenGLDriver.cpp#L2285-L2299

the new driver uses no matrices for 2D drawing at all:
https://github.com/minetest/minetest/blob/master/client/shaders/Irrlicht/Renderer2D.vsh

**Comment by appgurueu** [comment]
@GefullteTaubenbrust2 shared the following theory on Discord for why this might be happening:

> aw naah ive had this happen before, its the absolute worst
> i think opengl sometimes just shits itself when your geometry is exactly at the boundary between two pixels

and proposed the following solution:

> pretty sure way i fixed it when doing stuff with opengl is offset everything by half a pixel

So now the question is: Might we be somehow drawing exactly on boundaries? (I could imagine this happening for example if the screen height in pixels is an odd number and we're drawing a perfectly horizontally centered image with an even height.)

As for solutions, we should probably try to calculate pixel-perfect coordinates (this means rounding ourselves with a defined rounding mode), or we could probably just add some epsilon to force the GPU to round one way or another (assuming unspecified rounding mode is the problem here)...

**Comment by sfan5** [comment]
Soo do we consider this a blocker to using OpenGL 3 by default?

**Comment by Zughy** [comment]
Yes, please. We already have to deal with non-pixel perfect GUI/HUD elements, runining the quality of images.This makes it even worse

**Comment by ExeVirus** [comment]
Found a solution, probably the correct solution - full credit to Google's Gemini Pro for helping hunt it down. 

I'm on windows 10 x64, latest master as of this writing, debug build using MSVC and vcpkg toolchain. 

SDL, OpenGl 3 renderer, using the soothing32 texture pack, with minetest_game. 


Before and after: Notice the light 4-pixel "squares" in the middle left of each heart are correct in the bottom, but are being rounded all wrong in every heart above.

<img width="242" height="59" alt="Image" src="https://github.com/user-attachments/assets/fba383fd-248b-44a3-b849-4bf3feea26c0" />

The reason for this, based on testing and from Gemini, is because, in our Fixed Function pipeline, we have [this line of code](https://github.com/luanti-org/luanti/blob/c625fa71e108ecbf99a2ac5a99b7f59e20008978/irr/src/COpenGLDriver.cpp#L2266):
```c++
glTranslatef(0.375f, 0.375f, 0.0f);
```
For all 2d rendering: COpenGLDriver::setRenderStates2DMode()

The reason we do this makes sense our 2d renders are pixel perfect, so when I write to a pixel at 0,0 I'm actually rendering right on a pixel boundary:

<img width="134" height="143" alt="Image" src="https://github.com/user-attachments/assets/8e638778-ac6b-4ee0-b481-aac60462f5c1" />

Adding the 0.375f offset pushes the pixel firmly into the actual center of the 0,0 pixel (0.5,0.5):

<img width="127" height="167" alt="Image" src="https://github.com/user-attachments/assets/af2a2aa7-af3b-4a64-a490-6cee50e66796" />

For the OpenGl3 pipeline, this code change needed to happen in Driver.cpp in irr/src, specifically in the 

void COpenGL3DriverBase::draw2DImage()

Which is used for rendering HUD 2d images (hotbar, for example, among other textures)

[Convenient location to add the offset](https://github.com/luanti-org/luanti/blob/c625fa71e108ecbf99a2ac5a99b7f59e20008978/irr/src/OpenGL/Driver.cpp#L772):
```c++
	const f32 offset = 0.375f;
	f32 left  = (f32)destRect.UpperLeftCorner.X + offset;
	f32 right = (f32)destRect.LowerRightCorner.X + offset;
	f32 down  = (f32)destRect.LowerRightCorner.Y + offset;
	f32 top   = (f32)destRect.UpperLeftCorner.Y + offset;
```


TLDR:

In the fixed function pipeline, we setup openGL to render 2d images with an offset to that we render to the center of pixels and not the edge. Without that offset setup in shader-land, we must add that offset in C++ land calculations. No idea, but perhaps this is also a problem for formspecs or other 2d renderings?
